File: java/src/main/java/com/penapereira/patterns/observer/ConcreteObserver.java
package com.penapereira.patterns.observer;

import java.util.function.Consumer;

/** Reports the change it was told about through a callback supplied by its creator. */
public final class ConcreteObserver implements Observer {

    private final String name;
    private final Consumer<String> report;

    public ConcreteObserver(String name, Consumer<String> report) {
        this.name = name;
        this.report = report;
    }

    @Override
    public void update(int oldState, int newState) {
        report.accept(name + " notified: state " + oldState + " -> " + newState);
    }
}

File: javascript/src/observer/concrete-observer.js
/** Reports the change it was told about through a callback supplied by its creator. */
export class ConcreteObserver {
  #name;
  #report;

  constructor(name, report) {
    this.#name = name;
    this.#report = report;
  }

  update(oldState, newState) {
    this.#report(`${this.#name} notified: state ${oldState} -> ${newState}`);
  }
}
